diff --git a/misc/fuse2fs.c b/misc/fuse2fs.c
index c5957212..5f9d8976 100644
--- a/misc/fuse2fs.c
+++ b/misc/fuse2fs.c
@@ -329,6 +329,12 @@ struct fuse2fs {
 	unsigned int next_generation;
 };
 
+static inline struct fuse_context* fuse_get_sudo_context() {
+  struct fuse_context* ctxt = fuse_get_context();
+  ctxt->uid = 0;
+  return ctxt;
+}
+
 #define FUSE2FS_CHECK_MAGIC(fs, ptr, num) do {if ((ptr)->magic != (num)) \
 	return translate_error((fs), 0, EXT2_ET_MAGIC_EXT2_FILE); \
 } while (0)
@@ -631,7 +637,7 @@ static int fs_writeable(ext2_filsys fs)
 
 static int check_inum_access(ext2_filsys fs, ext2_ino_t ino, mode_t mask)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct ext2_inode inode;
 	mode_t perms;
@@ -701,7 +707,7 @@ static int check_inum_access(ext2_filsys fs, ext2_ino_t ino, mode_t mask)
 
 static void op_destroy(void *p EXT2FS_ATTR((unused)))
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -729,7 +735,7 @@ static void op_destroy(void *p EXT2FS_ATTR((unused)))
 
 static void *op_init(struct fuse_conn_info *conn)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -799,7 +805,7 @@ static int stat_inode(ext2_filsys fs, ext2_ino_t ino, struct stat *statbuf)
 
 static int op_getattr(const char *path, struct stat *statbuf)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t ino;
@@ -823,7 +829,7 @@ out:
 
 static int op_readlink(const char *path, char *buf, size_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -899,7 +905,7 @@ out:
 
 static int op_mknod(const char *path, mode_t mode, dev_t dev)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t parent, child;
@@ -1029,7 +1035,7 @@ out:
 
 static int op_mkdir(const char *path, mode_t mode)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t parent, child;
@@ -1296,7 +1302,7 @@ out:
 
 static int op_unlink(const char *path)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	int ret;
 
@@ -1406,7 +1412,7 @@ out:
 
 static int op_rmdir(const char *path)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	int ret;
 
@@ -1419,7 +1425,7 @@ static int op_rmdir(const char *path)
 
 static int op_symlink(const char *src, const char *dest)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t parent, child;
@@ -1543,7 +1549,7 @@ static int update_dotdot_helper(ext2_ino_t dir EXT2FS_ATTR((unused)),
 
 static int op_rename(const char *from, const char *to)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -1768,7 +1774,7 @@ out:
 
 static int op_link(const char *src, const char *dest)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	char *temp_path;
@@ -1872,7 +1878,7 @@ out:
 
 static int op_chmod(const char *path, mode_t mode)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -1931,7 +1937,7 @@ out:
 
 static int op_chown(const char *path, uid_t owner, gid_t group)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -2001,7 +2007,7 @@ out:
 
 static int op_truncate(const char *path, off_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -2141,7 +2147,7 @@ out:
 
 static int op_open(const char *path, struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	int ret;
 
@@ -2156,7 +2162,7 @@ static int op_read(const char *path EXT2FS_ATTR((unused)), char *buf,
 		   size_t len, off_t offset,
 		   struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -2213,7 +2219,7 @@ static int op_write(const char *path EXT2FS_ATTR((unused)),
 		    const char *buf, size_t len, off_t offset,
 		    struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -2285,7 +2291,7 @@ out:
 static int op_release(const char *path EXT2FS_ATTR((unused)),
 		      struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -2315,7 +2321,7 @@ static int op_fsync(const char *path EXT2FS_ATTR((unused)),
 		    int datasync EXT2FS_ATTR((unused)),
 		    struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -2342,7 +2348,7 @@ static int op_fsync(const char *path EXT2FS_ATTR((unused)),
 static int op_statfs(const char *path EXT2FS_ATTR((unused)),
 		     struct statvfs *buf)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	uint64_t fsid, *f;
@@ -2414,7 +2420,7 @@ static struct xattr_translate xattr_translators[] = {
 static int op_getxattr(const char *path, const char *key, char *value,
 		       size_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	struct ext2_xattr_handle *h;
@@ -2517,7 +2523,7 @@ static int copy_names(char *name, char *value EXT2FS_ATTR((unused)),
 
 static int op_listxattr(const char *path, char *names, size_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	struct ext2_xattr_handle *h;
@@ -2595,7 +2601,7 @@ static int op_setxattr(const char *path EXT2FS_ATTR((unused)),
 		       const char *key, const char *value,
 		       size_t len, int flags EXT2FS_ATTR((unused)))
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	struct ext2_xattr_handle *h;
@@ -2672,7 +2678,7 @@ out:
 
 static int op_removexattr(const char *path, const char *key)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	struct ext2_xattr_handle *h;
@@ -2763,7 +2769,7 @@ static int op_readdir(const char *path EXT2FS_ATTR((unused)),
 		      off_t offset EXT2FS_ATTR((unused)),
 		      struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -2797,7 +2803,7 @@ out:
 
 static int op_access(const char *path, int mask)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	errcode_t err;
@@ -2825,7 +2831,7 @@ out:
 
 static int op_create(const char *path, mode_t mode, struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t parent, child;
@@ -2952,7 +2958,7 @@ out:
 static int op_ftruncate(const char *path EXT2FS_ATTR((unused)),
 			off_t len, struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -3005,7 +3011,7 @@ static int op_fgetattr(const char *path EXT2FS_ATTR((unused)),
 		       struct stat *statbuf,
 		       struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	struct fuse2fs_file_handle *fh =
@@ -3025,7 +3031,7 @@ static int op_fgetattr(const char *path EXT2FS_ATTR((unused)),
 
 static int op_utimens(const char *path, const struct timespec ctv[2])
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct timespec tv[2];
 	ext2_filsys fs;
@@ -3117,7 +3123,7 @@ static int ioctl_setflags(ext2_filsys fs, struct fuse2fs_file_handle *fh,
 	struct ext2_inode_large inode;
 	int ret;
 	__u32 flags = *(__u32 *)data;
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 
 	FUSE2FS_CHECK_MAGIC(fs, fh, FUSE2FS_FILE_MAGIC);
@@ -3174,7 +3180,7 @@ static int ioctl_setversion(ext2_filsys fs, struct fuse2fs_file_handle *fh,
 	struct ext2_inode_large inode;
 	int ret;
 	__u32 generation = *(__u32 *)data;
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 
 	FUSE2FS_CHECK_MAGIC(fs, fh, FUSE2FS_FILE_MAGIC);
@@ -3262,7 +3268,7 @@ static int op_ioctl(const char *path EXT2FS_ATTR((unused)), int cmd,
 		    struct fuse_file_info *fp,
 		    unsigned int flags EXT2FS_ATTR((unused)), void *data)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -3305,7 +3311,7 @@ static int op_ioctl(const char *path EXT2FS_ATTR((unused)), int cmd,
 static int op_bmap(const char *path, size_t blocksize EXT2FS_ATTR((unused)),
 		   uint64_t *idx)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs;
 	ext2_ino_t ino;
@@ -3338,7 +3344,7 @@ out:
 static int fallocate_helper(struct fuse_file_info *fp, int mode, off_t offset,
 			    off_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -3474,7 +3480,7 @@ static errcode_t clean_block_edge(ext2_filsys fs, ext2_ino_t ino,
 static int punch_helper(struct fuse_file_info *fp, int mode, off_t offset,
 			off_t len)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	struct fuse2fs_file_handle *fh =
 		(struct fuse2fs_file_handle *)(uintptr_t)fp->fh;
@@ -3544,7 +3550,7 @@ static int op_fallocate(const char *path EXT2FS_ATTR((unused)), int mode,
 			off_t offset, off_t len,
 			struct fuse_file_info *fp)
 {
-	struct fuse_context *ctxt = fuse_get_context();
+	struct fuse_context *ctxt = fuse_get_sudo_context();
 	struct fuse2fs *ff = (struct fuse2fs *)ctxt->private_data;
 	ext2_filsys fs = ff->fs;
 	int ret;
